### ISSUE TITLE
Question: How do you bind a subclass in Gravity

### ISSUE THREAD

**Issue by untoldengine** [marcobambini/gravity#361]
Hi @marcobambini,

Sorry to bother you again, but I have another question regarding binding subclasses in Gravity.

Assume I have a class in my engine called

```
class Player{
//....
}
```

And I want to create a subclass of my Player class in a gravity script, for example:

```
//gravity script

class Defender:Player{
//...
}
```

How do you bind a subclass of type Player? Do you have any code examples, or can you point to where I should look?

Thanks again for your help.

**Comment by marcobambini** [comment]
Hi @untoldengine,

I am sorry but I am not sure to understand your question.
What do you mean by "How do you bind a subclass of type Player?"

**Comment by untoldengine** [comment]
Hi @marcobambini 

My question is basically, how do you bind derived classes with Gravity.

I'm trying to figure out how to do the following:

Let's say that I have a class in my engine called Entity.

```
//C++ class in game engine

class Entity{

 void init();
 void update();
}
```

I want to derive a class from Entity and add a new method 'applyVelocity()', but I want to do so in Gravity:

For example:

```
//gravity script

extern var Entity;

//deriving a subclass from the superclass Entity
class Hero:Entity{

    func applyVelocity(){...}

}
``` 

My question is, how do you bind derived classes with Gravity? I want to be able to call inherited behaviors and new behaviors in Gravity.

For example:

```
var hero = Hero(); //Hero is a derived class of Entity
hero.init(); //init is a method defined in the superclass Entity
hero.applyVelocity() //applyVelocity is defined in the Hero class
```

Hopefully, this clears up the confusion.

**Comment by marcobambini** [comment]
Hi @untoldengine,

Thanks a lot for the clarification.

I have to admit that even if it was technically possible, I never used to subclass an externally declared class.
There were some issues to fix (already fixed in the latest 0.8.2 version), so it should be safe to use now.

The example you described (class Hero subclass of extern class Entity) should now work without any special change.
This is the first implementation. For more complex cases, more work should probably be done.

Please let me know how it works.

**Comment by untoldengine** [comment]
Hey @marcobambini 

Thank you so much. It is working now. I think I have all that I need to start integrating Gravity into my engine (well, I hope :)).

My goal is to make Gravity the official scripting language for the Untold Engine. Internally, the game engine is written in C++ and uses the Apple Metal Graphics API for rendering. Even though you can develop a game using C++, I want to give game developers the flexibility to develop their games using a scripting language such as Gravity. I was looking into Lua but found the C++ bindings a bit confusing.

It should take me a couple of months to do all the required integrations (Gravity+Imgui+Untold Engine). 

I'll keep you posted.

Thanks for your help.

**Comment by marcobambini** [comment]
@untoldengine yes please keep me updated.
I'll do my best to support you in this very interesting project.

**Comment by untoldengine** [comment]
Hi @marcobambini 

Need your help one more time :) 

I'm getting the following error whenever I try to create an instance of a class inside the Hero class:

"Cannot set superclass of Hero to non class object."

Here is an example of what I'm doing:

```
extern var Entity;
extern var Vector3n;

class Hero:Entity{

	func fight(){
		var myVector=Vector3n();  //This is producing an error whenever I try to create an instance of Vector3n  
	}

}
```

I figured that the problem may be with the 'extern var Vector3n', so I did another test and declared a class in the same file, hoping that the Hero class would be able to see the declaration:


```
extern var Entity;

class MyClass{

	func foo(a,b){
		var c=a+b;
		return c;
	}
}

class Hero:Entity{

	func fight(){
		var myClass=MyClass(); //This is producing an error whenever I try to create an instance of Vector3n
	}

}

```

But I ended up getting the same error.

It seems that the problem is in the function 'gnode_t *lookup_identifier' in gravity_semacheck2.c. It seems that the symbol is null and the class declaration can't be found. Lines 174-179

Any ideas?

Thanks again for your help.
